src/main/kotlin/uk/gov/dluhc/printapi/service/PrintService.kt

```diff
@@ -3,20 +3,34 @@ package uk.gov.dluhc.printapi.service
 import org.springframework.stereotype.Service
 import uk.gov.dluhc.printapi.database.repository.CertificateRepository
 import uk.gov.dluhc.printapi.mapper.CertificateMapper
+import uk.gov.dluhc.printapi.mapper.SourceTypeMapper
 import uk.gov.dluhc.printapi.messaging.models.SendApplicationToPrintMessage
 import javax.transaction.Transactional
 import uk.gov.dluhc.printapi.client.ElectoralRegistrationOfficeManagementApiClient as EroClient
 
 @Service
 class PrintService(
     private val eroClient: EroClient,
+    private val sourceTypeMapper: SourceTypeMapper,
     private val certificateMapper: CertificateMapper,
     private val certificateRepository: CertificateRepository
 ) {
     @Transactional
     fun savePrintMessage(message: SendApplicationToPrintMessage) {
         val ero = eroClient.getEro(message.gssCode!!)
-        val certificate = certificateMapper.toCertificate(message, ero)
+
+        var certificate = certificateRepository.findByGssCodeInAndSourceTypeAndSourceReference(
+            gssCodes = listOf(message.gssCode),
+            sourceType = sourceTypeMapper.toSourceTypeEntity(message.sourceType),
+            sourceReference = message.sourceReference
+        )
+
+        if (certificate != null) {
+            certificateMapper.addPrintRequestToCertificate(message, ero, certificate)
+        } else {
+            certificate = certificateMapper.toCertificate(message, ero)
+        }
```

Comment:
Explored using Kotlin function scopes but looked more confusing than helpful.

Reply:
V Minor - If cert is null then create createCert else map it. (Instead of !=null)